cmd/clusterctl/client/cluster/mover.go

```diff
@@ -758,15 +774,18 @@ func (o *objectMover) ensureNamespace(toProxy Proxy, namespace string) error {
 }
 
 // createGroup creates all the Kubernetes objects into the target management cluster corresponding to the object graph nodes in a moveGroup.
-func (o *objectMover) createGroup(group moveGroup, toProxy Proxy) error {
+func (o *objectMover) createGroup(group moveGroup, toProxy Proxy, mutators ...ResourceMutatorFunc) error {
 	createTargetObjectBackoff := newWriteBackoff()
 	errList := []error{}
 
+	// Maintain a cache of namespaces that have been verified to already exist.
+	// Nb. This prevents us from making repetitive (and expensive) calls in listing all namespaces to ensure a namespace exists before creating a resource.
+	existingNamespaces := sets.New[string]()
```

Comment:
I don't think this should be managed in clusterctl, because it is a specific concern of the mutators you are looking to implement.

I think you can achieve the same by implementing your mutators as a method of a struct that maintains the list of existing namespaces, and this will give a cleaner and self contained solution totally transparent to clusterctl

Reply:
+1